C error when compiling `if a is string { for x in a {` where `a` has type `interface Any {}`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.3 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0

getwd: /v/cleanv
vmodules: /home/delian/.vmodules
vroot: /v/cleanv
vexe: /v/cleanv/v
vexe mtime: 2022-04-22 17:22:58
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 a2db44b.5dce091

Git version: git version 2.25.1
Git vroot status: weekly.2022.16-45-g365e7d6b
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 eee03626

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg any_interface_string_bug.v`
```v
interface Any {}

fn abc(a Any) {
	if a is string {
		for x in a {
			dump(x)
		}
	}
	if a is []u8 {
		for x in a {
			dump(x)
		}
	}
	if a is map[string]string {
		for k, v in a {
			println('> k: $k | v: $v')
		}
	}
	println(@FN)
}

fn main() {
	abc('abc')
	abc([u8(23), 77])
	abc({
		'abc': 'xyz'
		'def': 'jkl'
	})
	abc(123)
}

```

**What did you expect to see?**

A compiled and running program, that prints the bytes of the passed string.


**What did you see instead?**
```
==================
/tmp/v_1000/any_interface_string_bug.7438476477100425228.tmp.c:11560: error: struct or union expected
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

Note: with the following change, the above program works fine:
```
if a is string {
  s := *a
  for x in s {
```
printing:
```
[zz.v:7] x: 97
[zz.v:7] x: 98
[zz.v:7] x: 99
abc
[zz.v:12] x: 23
[zz.v:12] x: 77
abc
> k: abc | v: xyz
> k: def | v: jkl
abc
abc
```

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -95,6 +95,9 @@ fn (mut c Checker) for_in_stmt(mut node ast.ForInStmt) {
 				node.scope.update_var_type(node.key_var, key_type)
 			}
 			mut value_type := c.table.value_type(typ)
+			if sym.kind == .string {
+				value_type = ast.byte_type
+			}
 			if value_type == ast.void_type || typ.has_flag(.optional) {
 				if typ != ast.void_type {
 					c.error('for in: cannot index `${c.table.type_to_str(typ)}`', node.cond.pos())

--- a/vlib/v/gen/c/for.v
+++ b/vlib/v/gen/c/for.v
@@ -299,14 +299,15 @@ fn (mut g Gen) for_in_stmt(node ast.ForInStmt) {
 		} else {
 			node.cond
 		}
+		field_accessor := if node.cond_type.is_ptr() { '->' } else { '.' }
 		i := if node.key_var in ['', '_'] { g.new_tmp_var() } else { node.key_var }
 		g.write('for (int $i = 0; $i < ')
 		g.expr(cond)
-		g.writeln('.len; ++$i) {')
+		g.writeln('${field_accessor}len; ++$i) {')
 		if node.val_var != '_' {
-			g.write('\tbyte ${c_name(node.val_var)} = ')
+			g.write('\tu8 ${c_name(node.val_var)} = ')
 			g.expr(cond)
-			g.writeln('.str[$i];')
+			g.writeln('${field_accessor}str[$i];')
 		}
 	} else if node.kind == .struct_ {
 		cond_type_sym := g.table.sym(node.cond_type)

--- a/vlib/v/tests/inout/printing_for_v_in_a.out
+++ b/vlib/v/tests/inout/printing_for_v_in_a.out
@@ -1,0 +1,11 @@
+97
+98
+99
+abc
+23
+77
+abc
+> k: abc | v: xyz
+> k: def | v: jkl
+abc
+abc

--- a/vlib/v/tests/inout/printing_for_v_in_a.vv
+++ b/vlib/v/tests/inout/printing_for_v_in_a.vv
@@ -1,0 +1,30 @@
+interface Any {}
+
+fn abc(a Any) {
+	if a is string {
+		for x in a {
+			println(x)
+		}
+	}
+	if a is []u8 {
+		for x in a {
+			println(x)
+		}
+	}
+	if a is map[string]string {
+		for k, v in a {
+			println('> k: $k | v: $v')
+		}
+	}
+	println(@FN)
+}
+
+fn main() {
+	abc('abc')
+	abc([u8(23), 77])
+	abc({
+		'abc': 'xyz'
+		'def': 'jkl'
+	})
+	abc(123)
+}

